spring-integration-sftp: org.apache.sshd.common.io.WritePendingException: A write operation is already pending;

### Discussed in https://github.com/spring-projects/spring-integration/discussions/8707

<div type='discussions-op-text'>

<sup>Originally posted by **Ducius6** August 18, 2023</sup>
When we upgraded out application to Spring boot 3 we started getting issues like this: 

Caused by: org.apache.sshd.common.io.WritePendingException: A write operation is already pending; cannot write 21 bytes
	at org.apache.sshd.common.channel.ChannelAsyncOutputStream.writeBuffer(ChannelAsyncOutputStream.java:102)
	at org.apache.sshd.sftp.client.impl.DefaultSftpClient.send(DefaultSftpClient.java:299)
	at org.apache.sshd.sftp.client.impl.AbstractSftpClient.checkHandle(AbstractSftpClient.java:230)
	at org.apache.sshd.sftp.client.impl.AbstractSftpClient.openDir(AbstractSftpClient.java:829)
	at org.apache.sshd.sftp.client.impl.SftpDirEntryIterator.<init>(SftpDirEntryIterator.java:60)
	at org.apache.sshd.sftp.client.impl.SftpIterableDirEntry.iterator(SftpIterableDirEntry.java:64)
	at org.apache.sshd.sftp.client.impl.SftpIterableDirEntry.iterator(SftpIterableDirEntry.java:34)
	at java.base/java.lang.Iterable.spliterator(Iterable.java:101)
	at org.springframework.integration.sftp.session.SftpSession.doList(SftpSession.java:103)
	at org.springframework.integration.sftp.session.SftpSession.list(SftpSession.java:69)
	at org.springframework.integration.sftp.session.SftpSession.list(SftpSession.java:52)
	at org.springframework.integration.file.remote.session.CachingSessionFactory$CachedSession.list(CachingSessionFactory.java:227)
	at org.springframework.integration.file.remote.RemoteFileTemplate.lambda$list$5(RemoteFileTemplate.java:422)
	at org.springframework.integration.file.remote.RemoteFileTemplate.execute(RemoteFileTemplate.java:452)
	... 18 common frames omitted


We found out that issues is when you call **RemoteFileTemplate.list** method concurrently. Since ChannelAsyncOutputStream.writeBuffer cannot be caller concurrently. 

Before upgrading to SB3 we had no issues like this. 

So my question is, is this a bug in new implementation in SB3 using apache mina SSHD or is this expected behaviour and we should not call this method from different threads?</div>

--- a/spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/DefaultSftpSessionFactory.java
+++ b/spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/DefaultSftpSessionFactory.java
@@ -36,13 +36,15 @@ import org.apache.sshd.client.session.ClientSession;
 import org.apache.sshd.common.SshConstants;
 import org.apache.sshd.common.config.keys.FilePasswordProvider;
 import org.apache.sshd.common.keyprovider.KeyIdentityProvider;
+import org.apache.sshd.common.util.buffer.Buffer;
 import org.apache.sshd.common.util.io.resource.AbstractIoResource;
 import org.apache.sshd.common.util.io.resource.IoResource;
 import org.apache.sshd.common.util.net.SshdSocketAddress;
 import org.apache.sshd.common.util.security.SecurityUtils;
 import org.apache.sshd.sftp.client.SftpClient;
-import org.apache.sshd.sftp.client.SftpClientFactory;
+import org.apache.sshd.sftp.client.SftpErrorDataHandler;
 import org.apache.sshd.sftp.client.SftpVersionSelector;
+import org.apache.sshd.sftp.client.impl.DefaultSftpClient;
 
 import org.springframework.core.io.Resource;
 import org.springframework.integration.file.remote.session.SessionFactory;
@@ -281,8 +283,8 @@ public class DefaultSftpSessionFactory implements SessionFactory<SftpClient.DirE
 			boolean freshSftpClient = false;
 			if (sftpClient == null || !sftpClient.isOpen()) {
 				sftpClient =
-						SftpClientFactory.instance()
-								.createSftpClient(initClientSession(), this.sftpVersionSelector);
+						new ConcurrentSftpClient(initClientSession(), this.sftpVersionSelector,
+								SftpErrorDataHandler.EMPTY);
 				freshSftpClient = true;
 			}
 			sftpSession = new SftpSession(sftpClient);
@@ -393,6 +395,33 @@ public class DefaultSftpSessionFactory implements SessionFactory<SftpClient.DirE
 	public void resetSharedSession() {
 		Assert.state(this.isSharedSession, "Shared sessions are not being used");
 		this.sharedSftpClient = null;
+	}
+
+	/**
+	 * The {@link DefaultSftpClient} extension to lock the {@link #send(int, Buffer)}
+	 * for concurrent interaction.
+	 */
+	private static class ConcurrentSftpClient extends DefaultSftpClient {
+
+		private final Lock sendLock = new ReentrantLock();
+
+		ConcurrentSftpClient(ClientSession clientSession, SftpVersionSelector initialVersionSelector,
+				SftpErrorDataHandler errorDataHandler) throws IOException {
+
+			super(clientSession, initialVersionSelector, errorDataHandler);
+		}
+
+		@Override
+		public int send(int cmd, Buffer buffer) throws IOException {
+			this.sendLock.lock();
+			try {
+				return super.send(cmd, buffer);
+			}
+			finally {
+				this.sendLock.unlock();
+			}
+		}
+
 	}
 
 }

--- a/spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/SftpSession.java
+++ b/spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/SftpSession.java
@@ -22,8 +22,6 @@ import java.io.OutputStream;
 import java.io.UncheckedIOException;
 import java.net.SocketAddress;
 import java.time.Duration;
-import java.util.concurrent.locks.Lock;
-import java.util.concurrent.locks.ReentrantLock;
 import java.util.stream.Stream;
 import java.util.stream.StreamSupport;
 
@@ -49,12 +47,9 @@ import org.springframework.util.StringUtils;
  * @author Gary Russell
  * @author Artem Bilan
  * @author Christian Tzolov
- *
  * @since 2.0
  */
 public class SftpSession implements Session<SftpClient.DirEntry> {
-
-	private final Lock lock = new ReentrantLock();
 
 	private final SftpClient sftpClient;
 
@@ -113,7 +108,7 @@ public class SftpSession implements Session<SftpClient.DirEntry> {
 			}
 		}
 		remoteDir =
-				remoteDir.length() > 0 && remoteDir.charAt(0) == '/'
+				!remoteDir.isEmpty() && remoteDir.charAt(0) == '/'
 						? remoteDir
 						: this.sftpClient.canonicalPath(remoteDir);
 		return StreamSupport.stream(this.sftpClient.readDir(remoteDir).spliterator(), false)
@@ -138,30 +133,18 @@ public class SftpSession implements Session<SftpClient.DirEntry> {
 
 	@Override
 	public void write(InputStream inputStream, String destination) throws IOException {
-		this.lock.lock();
-		try {
-			OutputStream outputStream = this.sftpClient.write(destination);
-			FileCopyUtils.copy(inputStream, outputStream);
-		}
-		finally {
-			this.lock.unlock();
-		}
+		OutputStream outputStream = this.sftpClient.write(destination);
+		FileCopyUtils.copy(inputStream, outputStream);
 	}
 
 	@Override
 	public void append(InputStream inputStream, String destination) throws IOException {
-		this.lock.lock();
-		try {
-			OutputStream outputStream =
-					this.sftpClient.write(destination,
-							SftpClient.OpenMode.Create,
-							SftpClient.OpenMode.Write,
-							SftpClient.OpenMode.Append);
-			FileCopyUtils.copy(inputStream, outputStream);
-		}
-		finally {
-			this.lock.unlock();
-		}
+		OutputStream outputStream =
+				this.sftpClient.write(destination,
+						SftpClient.OpenMode.Create,
+						SftpClient.OpenMode.Write,
+						SftpClient.OpenMode.Append);
+		FileCopyUtils.copy(inputStream, outputStream);
 	}
 
 	@Override

--- a/spring-integration-sftp/src/test/java/org/springframework/integration/sftp/session/SftpSessionFactoryTests.java
+++ b/spring-integration-sftp/src/test/java/org/springframework/integration/sftp/session/SftpSessionFactoryTests.java
@@ -18,11 +18,13 @@ package org.springframework.integration.sftp.session;
 
 import java.io.File;
 import java.io.IOException;
+import java.io.UncheckedIOException;
 import java.net.ConnectException;
 import java.time.Duration;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
+import java.util.stream.IntStream;
 
 import org.apache.sshd.client.SshClient;
 import org.apache.sshd.client.auth.password.PasswordIdentityProvider;
@@ -30,6 +32,7 @@ import org.apache.sshd.client.keyverifier.AcceptAllServerKeyVerifier;
 import org.apache.sshd.common.SshException;
 import org.apache.sshd.server.SshServer;
 import org.apache.sshd.server.keyprovider.SimpleGeneratorHostKeyProvider;
+import org.apache.sshd.sftp.client.SftpClient;
 import org.apache.sshd.sftp.server.SftpSubsystemFactory;
 import org.junit.jupiter.api.Test;
 
@@ -45,7 +48,6 @@ import static org.awaitility.Awaitility.await;
  * @author Gary Russell
  * @author Artem Bilan
  * @author Auke Zaaiman
- *
  * @since 3.0.2
  */
 public class SftpSessionFactoryTests {
@@ -151,6 +153,42 @@ public class SftpSessionFactoryTests {
 			sftpSessionFactory.setUser("user");
 
 			assertThatNoException().isThrownBy(sftpSessionFactory::getSession);
+		}
+	}
+
+	@Test
+	void concurrentSessionListDoesntCauseFailure() throws IOException {
+		try (SshServer server = SshServer.setUpDefaultServer()) {
+			server.setPasswordAuthenticator((arg0, arg1, arg2) -> true);
+			server.setPort(0);
+			server.setKeyPairProvider(new SimpleGeneratorHostKeyProvider(new File("hostkey.ser").toPath()));
+			server.setSubsystemFactories(Collections.singletonList(new SftpSubsystemFactory()));
+			server.start();
+
+			DefaultSftpSessionFactory sftpSessionFactory = new DefaultSftpSessionFactory();
+			sftpSessionFactory.setHost("localhost");
+			sftpSessionFactory.setPort(server.getPort());
+			sftpSessionFactory.setUser("user");
+			sftpSessionFactory.setPassword("pass");
+			sftpSessionFactory.setAllowUnknownKeys(true);
+
+			SftpSession session = sftpSessionFactory.getSession();
+
+			List<SftpClient.DirEntry[]> dirEntries =
+					IntStream.range(0, 10)
+							.boxed()
+							.parallel()
+							.map(i -> {
+								try {
+									return session.list(".");
+								}
+								catch (IOException e) {
+									throw new UncheckedIOException(e);
+								}
+							})
+							.toList();
+
+			assertThat(dirEntries).hasSize(10);
 		}
 	}
 

